docs/library/rp2.pio.rst

```diff
@@ -0,0 +1,164 @@
+    - *sideset_init* configures the pins used side-setting. There can be at most 5.
+
+    The following parameters are used by default, but can be overriden in `state_machine.init()`:
+
+    - *in_shiftdir* is the default direction the ISR will shift, either `PIO.SHIFT_LEFT` or `PIO.SHIFT_RIGHT`.
+    - *out_shiftdir* is the default direction the OSR will shift, either `PIO.SHIFT_LEFT` or `PIO.SHIFT_RIGHT`.
+    - *push_thresh* is the threshold in bits before auto-push or conditional re-pushing is triggered.
+    - *pull_thresh* is the threshold in bits before auto-push or conditional re-pushing is triggered.
+
+    The remaining parameters are:
+
+    - *autopush* configures whether auto-push is enabled.
+    - *autopull* configures whether auto-pull is enabled.
+
+.. function:: asm_pio_encode(instr, sideset_count)
+
+    Assemble a single PIO instruction. You usually want to use `asm_pio()` instead.
+
+    >>> rp2.asm_pio_encode("set(0, 1)", 0)
+    57345
+
+.. class:: PIOASMError
+
+    This exception is raised from `asm_pio()` or `asm_pio_encode()` if there is an error assembling a PIO program.
+
+
+PIO class
+---------
+
+.. class:: PIO(id)
+
+    Gets the PIO instance numbered *id*. The RP2040 has two PIO instances, numbered 0 and 1.
+
+    Raises a ``ValueError`` if any other argument is provided.
+
+.. method:: pio.add_program(program)
+
+    Add the *program* to the instruction memory of this PIO instance.
+
+    The amount of memory available for programs on each PIO instance is limited. If there isn't enough space left in the PIO's program memory, you'll get ``OSError`` 12 ``ENOMEM``.
+
+.. method:: pio.remove_program([program])
+
+    Remove *program* from the instruction memory of this PIO instance.
+
+    If no program is provided, it removes all programs.
+
+    It is not an error to remove a program which has already been removed.
+
+.. method:: pio.state_machine(id, [program, ...])
+
+    Gets the state machine numbered *id*. On the RP2040, each PIO instance has four state machines, numbered 0 to 3.
+
+    Optionally initialize it with a *program*: see `state_machine.init`.
+
+    >>> rp2.PIO(1).state_machine(3)
+    StateMachine(7)
+
+.. method:: pio.irq(handler=None, trigger=IRQ_SM0|IRQ_SM1|IRQ_SM2|IRQ_SM3, hard=False)
+
+    Returns the IRQ object for this PIO instance.
+
+    MicroPython only uses IRQ 0 on each PIO instance. IRQ 1 is not available.
+
+    Optionally configure it.
+
+.. data:: PIO.IN_LOW
+          PIO.IN_HIGH
+          PIO.OUT_LOW
+          PIO.OUT_HIGH
+
+    These constants are used for the *out_init*, *set_init*, and *sideset_init* arguments to `asm_pio`.
+
+.. data:: PIO.SHIFT_LEFT
+          PIO.SHIFT_RIGHT
+
+    These constants are used for the *in_shiftdir* and *out_shiftdir* arguments to `asm_pio` or `state_machine.init`.
+
+.. data:: PIO.IRQ_SM0
+          PIO.IRQ_SM1
+          PIO.IRQ_SM2
+          PIO.IRQ_SM3
+
+    These constants are used for the *trigger* argument to `pio.irq`.
+
```

Comment:
I haven't included the `JOIN_TX` and `JOIN_RX` comments from #7093, since that PR isn't merged yet.